### ISSUE TITLE
GhostBatchNorm maybe can take less time

### ISSUE THREAD

**Issue by joseluismoreira** [dreamquark-ai/tabnet#243]
## Feature request

I have noticed that GBN implementation has a for loop, and in my training it takes more time specially when the virtual batch size is much smaller than the real one.

**What is the expected behavior?**

I hope get near constant time in the GBN, in order to increase a lot the real batch size

**What is motivation or use case for adding/changing the behavior?**

Improve the time spent in the GBN loop.

**How should this be implemented in your opinion?**

I have found a  vectorized implementation here: https://arxiv.org/pdf/2007.08554.pdf , but do not use the learnable parameters of the pytorch implementation... But it can be added easily.

I have tested already and improves a lot te time as expected...

![image](https://user-images.githubusercontent.com/43185535/100030381-2578fc00-2dd2-11eb-9078-5da27aad6055.png)

**Are you willing to work on this yourself?**
yes

**Comment by joseluismoreira** [comment]
A problem with the code above is that the real batch should be divisible by the virtual batch. I dont know how to solve this yet..

**Comment by Optimox** [comment]
I think it's alright to force a virtual batch size to divide the batch size, we could also set drop_last=False systematically.

But what happens for prediction? That is a bit more complicated, we way artificially augment the last batch if it's not full and drop the lines before returning to the final user.

It looks like an interesting speed up!

**Comment by joseluismoreira** [comment]
I think it is not necessary, @Optimox, since in the test time, batch norm should use global statistics, so do ghostnorm.... btw, we have to implement this global statistics capturing, when we implement ghostnorm in a OOP way.  See the discussion here for more details about test time. https://github.com/tensorflow/tensorflow/issues/23050

**Comment by joseluismoreira** [comment]
May I implement it and send to you?

**Comment by eduardocarvp** [comment]
Hi @joseluismoreira,

Sure, you can implement your idea on a new branch. Like you said on your last comment we have to store the `running_mean` and `running_var` and implement as a PyTorch Module so that it behaves as a custom layer. I'm pretty sure we can do better that what exists today on the code.

If you create a PR we can continue the discussions there.

**Comment by Optimox** [comment]
@joseluismoreira I think the important part in the end is whether the total training time goes down, I think that currently this might not be a bottle neck, so it would be nice to see training time improvement in your PR
